plugins/ui-update/notifications.ts

```diff
@@ -0,0 +1,192 @@
+import { i18nMark } from "@lingui/react";
+import * as semver from "semver";
+import { filter } from "rxjs/operators";
+import { NotificationService } from "@extension-kid/notification-service";
+import {
+  ToastAppearance,
+  ToastCallbackType,
+  ToastNotification
+} from "@extension-kid/toast-notifications";
+
+import { getAction$, getUiMetadata$ } from "./streams";
+import { UIActions, UIActionType } from "./types/UIAction";
+import { rollbackUI, updateUI } from "./commands";
+
+const LOCAL_DEV_VERSION = "0.0.0";
+const activeNotifications: string[] = [];
+
+function clearActiveNotifications() {
+  if (activeNotifications.length) {
+    activeNotifications.splice(0, activeNotifications.length);
+  }
+}
+
+function removeActiveNotification(id: string) {
+  const notificationIndex = activeNotifications.indexOf(id);
+  if (notificationIndex !== -1) {
+    activeNotifications.splice(notificationIndex, 1);
+  }
+}
+
+function displayNotification(
+  notificationService: NotificationService,
+  notification: ToastNotification
+) {
+  if (!activeNotifications.includes(notification.id)) {
+    notificationService.push(notification);
+    activeNotifications.push(notification.id);
+  }
+}
+
+function updateAvailableCallback(
+  callbackType: ToastCallbackType,
+  toastNotification: ToastNotification
+) {
+  removeActiveNotification(toastNotification.id);
+  switch (callbackType) {
+    case "primary":
+      // Reload the UI
+      location.reload();
+      break;
+  }
+}
+
+function updateFailedCallback(
+  callbackType: ToastCallbackType,
+  toastNotification: ToastNotification,
+  updateVersion: string
+) {
+  removeActiveNotification(toastNotification.id);
+  switch (callbackType) {
+    case "primary":
+      // Retry update
+      updateUI(updateVersion);
+      break;
+  }
+}
+
+function rollbackFailedCallback(
+  callbackType: ToastCallbackType,
+  toastNotification: ToastNotification
+) {
+  removeActiveNotification(toastNotification.id);
+  switch (callbackType) {
+    case "primary":
+      // Retry rollback
+      rollbackUI();
+      break;
+  }
+}
+
+function dismissNotificationCallback(
+  _callbackType: ToastCallbackType,
+  toastNotification: ToastNotification
+) {
+  removeActiveNotification(toastNotification.id);
+}
+
+function setupUIUpdatedNotification(notificationService: NotificationService) {
+  getUiMetadata$().subscribe(uiMetadata => {
+    const { clientBuild, serverBuild } = uiMetadata;
+    const coercedClientBuild = semver.coerce(clientBuild || "");
+    const coercedServerBuild = semver.coerce(serverBuild || "");
+    if (
```

Comment:
here we have the logic under a conditional with no alternative `else` branch. It looks more like `filter` operator could be leveraged